### PR TITLE
feature: NoUnneededControlParenthesesFixer - Fix more cases

### PR DESCRIPTION
**Pull request by SpacePossum**
some of the cases raised here https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4997 , but not all

also closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5573

### CODE REVIEW

File: tests/Fixer/ControlStructure/NoUnneededControlParenthesesFixerTest.php [hunk 1]
```diff
@@ -430,125 +344,1221 @@ function foo() { $a = (yield($x)); }
...
+        ];
+
+        foreach ($assignOperators as $assignOperator) {
+            yield [
+                '<?php $a '.$assignOperator.' $x;',
+                '<?php $a '.$assignOperator.' ($x);',
+            ];
+        }
+
+        yield 'do not fix' => [
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply tests/Fixer/ControlStructure/NoUnneededControlParenthesesFixerTest.php:1188]: I've put these in the `do not fix` because I think these cases should not ever be fixed.
While we know that these `(` are currently `)` overcomplete, there is the explicit no BC promise from PHP to consider here I think:
```
Operator precedence and associativity only determine how expressions are grouped, they do not specify an order of evaluation. PHP does not (in the general case) specify in which order an expression is evaluated and code that assumes a specific order of evaluation should be avoided, because the behavior can change between versions of PHP or depending on the surrounding code.
```

https://www.php.net/manual/en/language.operators.precedence.php

I think this can be discussed in a follow up PR if someone really wants this to change :+1:

**Comment by SpacePossum** [comment]
rebased on master to resolve conflict, no other changes are made

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 2]
```diff
@@ -34,32 +35,42 @@
...
-    public function isCandidate(Tokens $tokens): bool
-    {
-        $types = [];
-
-        foreach (self::$loops as $loop) {
-            $types[] = (array) $loop['lookupTokens'];
-        }
-
-        $types = array_merge(...$types);
+    private static array $blockTypes = [
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php]: thanks, done

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 3]
```diff
@@ -167,20 +206,490 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+
+        $defaults = array_filter(
+            $options,
+            static function (string $option): bool {
+                return 'negative_instanceof' !== $option && 'others' !== $option && 'yield_from' !== $option;
+            }
+        );
+
         return new FixerConfigurationResolver([
             (new FixerOptionBuilder('statements', 'List of control statements to fix.'))
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:230]: I'm not sure if I would in favor of that. When testing these changes on a bunch of projects I found that some like negative `instanceof` with`(`  (like `!($foo instanceof $b);`) and some don't (like `!$foo instanceof $b;`). Some other much debated are `echo (1);` etc.

We could on next major see if all options still make sense, I agree that `case`, `break`, were added for BC reasons and might no longer make sense to keep as option.

File: doc/rules/control_structure/no_unneeded_control_parentheses.rst [hunk 4]
```diff
@@ -75,9 +72,9 @@ The rule is part of the following rule sets:
 @PhpCsFixer
   Using the `@PhpCsFixer <./../../ruleSets/PhpCsFixer.rst>`_ rule set will enable the ``no_unneeded_control_parentheses`` rule with the config below:
 
-  ``['statements' => ['break', 'clone', 'continue', 'echo_print', 'return', 'switch_case', 'yield', 'yield_from']]``
+  ``['statements' => ['break', 'clone', 'continue', 'echo_print', 'others', 'return', 'switch_case', 'yield', 'yield_from']]``
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply doc/rules/control_structure/no_unneeded_control_parentheses.rst]: done

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 5]
```diff
@@ -167,20 +206,490 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
      */
     protected function createConfigurationDefinition(): FixerConfigurationResolverInterface
     {
+        $options = [
+            'break',
+            'clone',
+            'continue',
+            'echo_print',
+            'negative_instanceof',
+            'others',
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:215]: Any other unneeded `(` `)` will be removed that is not covered by the other cases, which also allows for future extension without some strict BC promise.

File: doc/rules/control_structure/no_unneeded_control_parentheses.rst [hunk 6]
```diff
@@ -75,9 +72,9 @@ The rule is part of the following rule sets:
...
 
 @Symfony
   Using the `@Symfony <./../../ruleSets/Symfony.rst>`_ rule set will enable the ``no_unneeded_control_parentheses`` rule with the config below:
 
-  ``['statements' => ['break', 'clone', 'continue', 'echo_print', 'return', 'switch_case', 'yield', 'yield_from']]``
+  ``['statements' => ['break', 'clone', 'continue', 'echo_print', 'others', 'return', 'switch_case', 'yield', 'yield_from']]``
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply doc/rules/control_structure/no_unneeded_control_parentheses.rst:80]: There were only few hits on the SF repo (and other repo's of the same ecoculture like Twig), but good to have another check.

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 7]
```diff
@@ -101,63 +109,94 @@ public function getDefinition(): FixerDefinitionInterface
...
-            $prevToken = $tokens[$index];
+            $beforeOpenIndex = $tokens->getPrevMeaningfulToken($openIndex);
+            $afterCloseIndex = $tokens->getNextMeaningfulToken($closeIndex);
 
-            foreach ($loops as $loop) {
-                if (!$prevToken->isGivenKind($loop['lookupTokens'])) {
-                    continue;
+            // do a cheap check for negative case: `X()`
+
+            if ($tokens[$tokens->getNextMeaningfulToken($openIndex)]->equals(')')) {
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php]: thanks, done!

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 8]
```diff
@@ -167,20 +206,490 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+        return $tokens[$beforeOpenIndex]->isGivenKind(T_CLONE);
+    }
+
+    private function isWrappedCloneArgument(Tokens $tokens, int $beforeOpenIndex, int $openIndex, int $closeIndex, int $afterCloseIndex): bool
+    {
+        $beforeOpenIndex = $tokens->getPrevMeaningfulToken($beforeOpenIndex);
+
+        if (
+            !(
+                $tokens[$beforeOpenIndex]->equals('?') // For BC reasons
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:265]: TBH I'm not 100% convinced the current `clone` fixing is handling `?` correctly all the time. However, I didn't want to touch or change this behavior so I left it "as is".
I wouldn't drop it in the next major unless someone brings it up out of bug report or some other need to touch this code.

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 9]
```diff
@@ -101,63 +109,94 @@ public function getDefinition(): FixerDefinitionInterface
...
+            $beforeOpenIndex = $tokens->getPrevMeaningfulToken($openIndex);
+            $afterCloseIndex = $tokens->getNextMeaningfulToken($closeIndex);
 
-            foreach ($loops as $loop) {
-                if (!$prevToken->isGivenKind($loop['lookupTokens'])) {
-                    continue;
+            // do a cheap check for negative case: `X()`
+
+            if ($tokens[$tokens->getNextMeaningfulToken($openIndex)]->equals(')')) {
+                if ($this->isExitStatement($tokens, $beforeOpenIndex)) {
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:149]: I prefer the readability here, first I did most checks inline with comments, but still it is than pretty hard to follow.

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 10]
```diff
@@ -101,63 +109,94 @@ public function getDefinition(): FixerDefinitionInterface
...
+
+                continue;
+            }
 
-                $blockEndNextIndex = $tokens->getNextMeaningfulToken($blockEndIndex);
+            // handle `clone` statements
 
-                if (!$tokens[$blockEndNextIndex]->equalsAny($loop['neededSuccessors'])) {
-                    continue;
+            if ($this->isCloneStatement($tokens, $beforeOpenIndex)) {
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:172]: I prefer the readability here, first I did most checks inline with comments, but still it is than pretty hard to follow.

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 11]
```diff
@@ -167,20 +206,490 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+                $knownNegativeTypes[] = T_MATCH;
+            }
+        }
+
+        return $token->equalsAny($knownNegativeTypes);
+    }
+
+    private function containsOperation(Tokens $tokens, int $startIndex, int $endIndex): bool
+    {
+        static $noopTypes;
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:570]: I prefer it not to be, there is no (and should no be a) case of this list to be used outside of the function. Scoping it down here is more favorable for that reason IMHO.

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 12]
```diff
@@ -101,63 +109,94 @@ public function getDefinition(): FixerDefinitionInterface
...
+
+            $instanceOfIndex = $this->getIndexOfInstanceOfStatement($tokens, $openIndex, $closeIndex);
+
+            if (null !== $instanceOfIndex) {
+                if ($this->isWrappedInstanceOf($tokens, $instanceOfIndex, $beforeOpenIndex, $openIndex, $closeIndex, $afterCloseIndex)) {
+                    if ($tokens[$beforeOpenIndex]->equals('!')) {
+                        $this->removeUselessParenthesisPair($tokens, $beforeOpenIndex, $afterCloseIndex, $openIndex, $closeIndex, 'negative_instanceof');
+                    } else {
+                        $this->removeUselessParenthesisPair($tokens, $beforeOpenIndex, $afterCloseIndex, $openIndex, $closeIndex, 'others');
                     }
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php]: thanks, done!

File: tests/Fixer/ControlStructure/NoUnneededControlParenthesesFixerTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrmajor) [reply tests/Fixer/ControlStructure/NoUnneededControlParenthesesFixerTest.php:1188]: > Operator precedence and associativity only determine how expressions are grouped, they do not specify an order of evaluation.

What does that mean? Aren't they grouped in order to be evaluated according to these groupings?

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 13]
```diff
@@ -167,20 +206,490 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+        }
+
+        return null !== $forCandidateIndex && $tokens[$forCandidateIndex]->isGivenKind(T_FOR);
+    }
+
+    // `fn() => (X);`
+    private function isWrappedFnBody(Tokens $tokens, int $beforeOpenIndex, int $afterCloseIndex): bool
+    {
+        if (
+            !$tokens[$afterCloseIndex]->equalsAny([';', [T_CLOSE_TAG]])
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php]: thanks, more tests are added :+1:

File: src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php [hunk 14]
```diff
@@ -167,20 +206,490 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
...
+            || !$tokens[$beforeOpenIndex]->isGivenKind(T_DOUBLE_ARROW)
+        ) {
+            return false;
+        }
+
+        $beforeOpenIndex = $tokens->getPrevMeaningfulToken($beforeOpenIndex);
+
+        if (!$tokens[$beforeOpenIndex]->equals(')')) {
```
[the remark this answers is not in the document]
Comment (SpacePossum) [reply src/Fixer/ControlStructure/NoUnneededControlParenthesesFixer.php:506]: thanks so much, I've added support for this now

File: doc/rules/control_structure/no_unneeded_control_parentheses.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (SpacePossum) [reply doc/rules/control_structure/no_unneeded_control_parentheses.rst:80]: https://github.com/symfony/symfony/pull/46886
this demo's the effect of the proposed config change

**Review by julienfalque — APPROVED** [review]

**Comment by julienfalque** [comment]
Thank you @SpacePossum!
